public/app/features/plugins/plugin_loader.ts

```diff
@@ -74,7 +74,22 @@ export async function importPluginModule({
     return importPluginModuleInSandbox({ pluginId });
   }
 
-  return SystemJS.import(modulePath);
+  let mod = await SystemJS.import(modulePath);
+  let attempts = 0;
+  // While the module's default export is a promise, wait for it to resolve.
+  // Do we need to recurse here? It seems unlikely that a module would return a promise that resolves to another promise.
+  while (
+    !('plugin' in mod) &&
+    'default' in mod &&
+    typeof mod.default === 'object' &&
+    'then' in mod.default &&
+    typeof mod.default.then === 'function' &&
+    attempts < 5
+  ) {
+    mod = await mod.default;
```

Comment:
Promises don't have a timeout, why are we trying to re-attempt here?

Reply:
This was in an effort to avoid recursing infinitely into modules where `mod.default` is a promise all the way down. Honestly I'm not sure the while loop is even necessary, we could probably just handle it with an `if` and keep the current behaviour if there is more than one level of promise?

Reply:
E.g. something like this might be simpler, and works for my use case at least:
```js
  let mod = await SystemJS.import(modulePath);
  if (
    !('plugin' in mod) &&
    'default' in mod &&
    typeof mod.default === 'object' &&
    'then' in mod.default &&
    typeof mod.default.then === 'function'
  ) {
    mod = await mod.default;
  }
  return mod;
}
```